src/ruby/ext/grpc/rb_compression_options.c

```diff
@@ -0,0 +1,351 @@
+ * to the
+ * the GRPC core as a channel argument value.
+ * A nil return value means that it hasn't been set. */
+VALUE grpc_rb_compression_options_default_algorithm_internal_value(VALUE self) {
+  grpc_rb_compression_options *wrapper = NULL;
+
+  TypedData_Get_Struct(self, grpc_rb_compression_options,
+                       &grpc_rb_compression_options_data_type, wrapper);
+
+  if (wrapper->wrapped->default_algorithm.is_set) {
+    return INT2NUM(wrapper->wrapped->default_algorithm.algorithm);
+  } else {
+    return Qnil;
+  }
+}
+
+/* Gets the internal value of the default compression level that is to be passed
+ * to the GRPC core as a channel argument value.
+ * A nil return value means that it hasn't been set. */
+VALUE grpc_rb_compression_options_default_level_internal_value(VALUE self) {
+  grpc_rb_compression_options *wrapper = NULL;
+
+  TypedData_Get_Struct(self, grpc_rb_compression_options,
+                       &grpc_rb_compression_options_data_type, wrapper);
+
+  if (wrapper->wrapped->default_level.is_set) {
+    return INT2NUM((int)wrapper->wrapped->default_level.level);
+  } else {
+    return Qnil;
+  }
+}
+
+/* Disables compression algorithms by their names. Raises an error if an unkown
+ * name was passed. */
+VALUE grpc_rb_compression_options_disable_algorithms(int argc, VALUE *argv,
+                                                     VALUE self) {
+  VALUE algorithm_names = Qnil;
+  VALUE ruby_str = Qnil;
+  grpc_compression_algorithm internal_algorithm_value;
+
+  /* read variadic argument list of names into the algorithm_name ruby array. */
+  rb_scan_args(argc, argv, "0*", &algorithm_names);
+
+  for (int i = 0; i < RARRAY_LEN(algorithm_names); i++) {
+    ruby_str =
+        rb_funcall(rb_ary_entry(algorithm_names, i), rb_intern("to_s"), 0);
+    grpc_rb_compression_options_get_internal_value_of_algorithm(
+        ruby_str, &internal_algorithm_value);
+    rb_funcall(self, rb_intern("disable_algorithm_internal"), 1,
+               LONG2NUM((long)internal_algorithm_value));
+  }
+
+  return Qnil;
+}
+
+/* Provides a ruby hash of GRPC core channel argument key-values that
+ * correspond to the compression settings on this instance. */
+VALUE grpc_rb_compression_options_to_hash(VALUE self) {
+  grpc_rb_compression_options *wrapper = NULL;
+  grpc_compression_options *compression_options = NULL;
+  VALUE channel_arg_hash = rb_funcall(rb_cHash, rb_intern("new"), 0);
+
+  TypedData_Get_Struct(self, grpc_rb_compression_options,
+                       &grpc_rb_compression_options_data_type, wrapper);
+  compression_options = wrapper->wrapped;
+
+  /* Add key-value pairs to the new Ruby hash. It can be used
+   * as GRPC core channel arguments. */
+  if (compression_options->default_level.is_set) {
+    rb_funcall(channel_arg_hash, rb_intern("[]="), 2,
+               rb_str_new2(GRPC_COMPRESSION_CHANNEL_DEFAULT_LEVEL),
+               INT2NUM((int)compression_options->default_level.level));
```

Comment:
This should be done with `rb_hash_aset`, like we use at https://github.com/grpc/grpc/blob/master/src/ruby/ext/grpc/rb_call.c#L483.